pkg/cli/admin/release/extract_tools.go

```diff
@@ -622,88 +623,113 @@ func (o *ExtractOptions) extractCommand(command string) error {
+		}
+
+		if len(replacement.marker) > longestMarker {
+			longestMarker = len(replacement.marker)
+			if bufferSize < longestMarker {
+				return fmt.Errorf("the buffer size must be greater than %d bytes to find %s", len(replacement.marker), replacement.name)
+			}
+		}
+	}
 
-func copyReplace(w io.Writer, r io.Reader, max int, releaseInfo, marker string) (bool, error) {
-	match := []byte(marker[:len(releaseInfo)+1])
 	offset := 0
-	buf := make([]byte, max+offset)
-	matched := false
+	buf := make([]byte, bufferSize)
+	remaining := len(replacements)
+	matches := make([]bool, remaining)
 
 	for {
 		n, err := io.ReadFull(r, buf[offset:])
 
 		// search in the buffer for the expected match
 		end := offset + n
 		if n > 0 {
-			index := bytes.Index(buf[:end], match)
-			if index != -1 {
-				klog.V(2).Infof("Found match at %d (len=%d, offset=%d, n=%d)", index, len(buf), offset, n)
-				// the replacement starts at the beginning of the match, contains the release string and a terminating NUL byte
-				copy(buf[index:index+len(releaseInfo)], []byte(releaseInfo))
-				buf[index+len(releaseInfo)] = 0x00
-				matched = true
+			for i, matched := range matches {
+				if !matched {
+					replacement := replacements[i]
+					index := bytes.Index(buf[:end], replacement.marker)
+					if index != -1 {
+						klog.V(2).Infof("Found match for %s at %d (len=%d, offset=%d, n=%d)", replacements[i].name, index, len(buf), offset, n)
+						// the replacement starts at the beginning of the match, contains the replacement value and a terminating NUL byte
+						copy(buf[index:index+len(replacement.value)], []byte(replacement.value))
+						buf[index+len(replacement.value)] = 0x00
+						matches[i] = true
+						remaining -= 1
+					}
+				}
 			}
 		}
 
 		// write everything that we have already searched (excluding the end of the buffer that will
 		// be checked next pass)
-		nextOffset := end - len(marker)
-		if nextOffset < 0 || matched {
-			nextOffset = 0
-		}
-		_, wErr := w.Write(buf[:end-nextOffset])
-		if wErr != nil {
-			return matched, wErr
+		writeTo := end - longestMarker
+		if writeTo < 0 {
+			writeTo = 0
+		}
+		if remaining == 0 || err != nil {
+			writeTo = end
+		}
+		offset = 0
+		for offset < writeTo {
+			n, wErr := w.Write(buf[offset:writeTo])
+			if wErr != nil {
+				return wErr
+			}
+			offset += n
 		}
+
 		if err != nil {
 			if err == io.EOF || err == io.ErrUnexpectedEOF {
-				return matched, nil
+				remainingNames := make([]string, 0, len(replacements))
+				for i, matched := range matches {
+					if !matched {
+						remainingNames = append(remainingNames, replacements[i].name)
+					}
+				}
+				sort.Strings(remainingNames)
+				if len(remainingNames) > 0 {
+					klog.Warningf("Unable to make all expected replacements in %s.  Remaining: %s", name, strings.Join(remainingNames, ", "))
```

Comment:
Hmm.  I see a number of these in master, but I also see things like `fmt.Fprintf(o.ErrOut, "warning...`.  What is the recommended approach to writing warnings from these CLI packages?

Reply:
In these commands warnings are printed with fprintf.

Klog is for debugging and servers, not end users.

Reply:
> In these commands warnings are printed with fprintf.

Seems like a pain to pass around an error-output `Writer`, but :man_shrugging:, updated to use `Fprintf` with 1604d6d28 -> cbcc1fd6c.